File: src/Practices/Array/ValidAnagram.ts
// leetcode: 242. Valid Anagram https://leetcode.com/problems/valid-anagram/

function isAnagram(s: string, t: string): boolean {
    const hashmap1 = {};

    for (let char of s)
        if (!hashmap1[char])
            hashmap1[char] = 1
        else
            hashmap1[char] += 1;


    for (let char of t)
        if (!hashmap1[char])
            return false
        else
            hashmap1[char] -= 1;


    for (let val of Object.values(hashmap1))
        if (val !== 0)
            return false

    return true;
};

console.log(isAnagram("anagram", "nagaram"));

File: src/Array n String/String/Basic onces/ValidAnagram.cpp
// Leetcode: 242. Valid Anagram (https://leetcode.com/problems/valid-anagram/)
// Tags: #easy
// Efficiency: Time Complexity O(n+m), Space Complexity O(1)

#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    bool isAnagram(string s, string t)
    {
        unordered_map<char, int> map;

        int s_size = s.length();
        int t_size = t.length();

        for (int i = 0; i < s_size; i++)
        {
            map[s[i]]++;
        }

        for (int i = 0; i < t_size; i++)
        {
            if (--map[t[i]] < 0)
            {
                return false;
            }
        }

        for (auto x : map)
        {
            if (x.second > 0)
            {
                return false;
            }
        }

        return true;
    }
};

int main()
{
    Solution solution;
    string s = "anagram";
    string t = "nagaram";
    bool output;

    output = solution.isAnagram(s, t);

    cout << output;
}